File: hotcold.js
// hotcold.js - Игра "Горячо-холодно" на JavaScript (Node.js)
const readline = require('readline').createInterface({
    input: process.stdin,
    output: process.stdout
});
const fs = require('fs').promises;

class Game {
    constructor(difficulty) {
        this.difficulty = difficulty;
        const ranges = {
            1: { low: 1, high: 50, maxAttempts: Infinity },
            2: { low: 1, high: 100, maxAttempts: 10 },
            3: { low: 1, high: 200, maxAttempts: 7 }
        };
        const { low, high, maxAttempts } = ranges[difficulty] || ranges[2];
        this.low = low;
        this.high = high;
        this.maxAttempts = maxAttempts;
        this.secret = Math.floor(Math.random() * (high - low + 1)) + low;
        this.attempts = 0;
        this.history = [];
        this.won = false;
        this.record = null;
        this.loadRecord();
    }

    checkGuess(guess) {
        if (guess === this.secret) {
            this.won = true;
            return "ПОБЕДА!";
        }
        const dist = Math.abs(guess - this.secret);
        let hint = dist <= 10 ? "ГОРЯЧО" : dist <= 25 ? "ТЕПЛО" : "ХОЛОДНО";
        hint += guess < this.secret ? " ▲" : " ▼";
        return hint;
    }

    addAttempt(guess, hint) {
        this.attempts++;
        this.history.push({ guess, hint });
    }

    isGameOver() {
        return this.won || this.attempts >= this.maxAttempts;
    }

    async saveRecord() {
        let records = {};
        try {
            const data = await fs.readFile('records.json', 'utf8');
            records = JSON.parse(data);
        } catch {}
        const key = `${this.difficulty}_${this.low}_${this.high}`;
        if (!records[key] || this.attempts < records[key]) {
            records[key] = this.attempts;
            await fs.writeFile('records.json', JSON.stringify(records, null, 2));
            return true;
        }
        return false;
    }

    async loadRecord() {
        try {
            const data = await fs.readFile('records.json', 'utf8');
            const records = JSON.parse(data);
            const key = `${this.difficulty}_${this.low}_${this.high}`;
            this.record = records[key] || null;
        } catch {}
    }

    displayHistory() {
        console.log('\x1b[36m--- История попыток ---\x1b[0m');
        this.history.forEach((h, i) => {
            const color = h.hint.includes('ГОРЯЧО') ? '\x1b[31m' : '\x1b[34m';
            console.log(`${i+1}: ${h.guess} -> ${color}${h.hint}\x1b[0m`);
        });
        console.log('------------------------');
    }
}

function askQuestion(query) {
    return new Promise(resolve => readline.question(query, resolve));
}

async function playGame() {
    console.log('\x1b[33m🔥❄️ Добро пожаловать в Hot&Cold!\x1b[0m');
    while (true) {
        console.log('\nВыберите уровень сложности:');
        console.log('1. Лёгкий (1-50, безлимит)');
        console.log('2. Средний (1-100, 10 попыток)');
        console.log('3. Сложный (1-200, 7 попыток)');
        console.log('4. Выйти');
        const choice = await askQuestion('Ваш выбор: ');
        if (choice === '4') {
            console.log('До свидания!');
            readline.close();
            process.exit(0);
        }
        if (!['1','2','3'].includes(choice)) {
            console.log('\x1b[31mНеверный выбор.\x1b[0m');
            continue;
        }
        const diff = parseInt(choice);
        const game = new Game(diff);
        await game.loadRecord();

        if (game.record !== null) {
            console.log(`\x1b[35mТекущий рекорд: ${game.record} попыток\x1b[0m`);
        } else {
            console.log('Рекорда пока нет.');
        }
        console.log(`\x1b[32mЧисло загадано! Диапазон ${game.low}-${game.high}\x1b[0m`);
        if (game.maxAttempts < Infinity) {
            console.log(`У вас ${game.maxAttempts} попыток.`);
        } else {
            console.log('Попытки не ограничены.');
        }

        while (!game.isGameOver()) {
            const input = await askQuestion(`Попытка ${game.attempts+1}: Введите число: `);
            const guess = parseInt(input);
            if (isNaN(guess)) {
                console.log('\x1b[31mВведите целое число!\x1b[0m');
                continue;
            }
            if (guess < game.low || guess > game.high) {
                console.log(`\x1b[31mЧисло должно быть от ${game.low} до ${game.high}\x1b[0m`);
                continue;
            }
            const hint = game.checkGuess(guess);
            game.addAttempt(guess, hint);
            let color = '\x1b[31m';
            if (hint.includes('ХОЛОДНО')) color = '\x1b[34m';
            else if (hint.includes('ПОБЕДА')) color = '\x1b[32m';
            console.log(`${color}${hint}\x1b[0m`);

            if (game.won) {
                console.log(`\x1b[32m🎉 ПОБЕДА! Число ${game.secret} угадано за ${game.attempts} попыток!\x1b[0m`);
                const newRecord = await game.saveRecord();
                if (newRecord) console.log('\x1b[36m🏆 Новый рекорд!\x1b[0m');
                break;
            } else if (game.attempts >= game.maxAttempts) {
                console.log(`\x1b[31m😞 Вы исчерпали все попытки. Загаданное число: ${game.secret}\x1b[0m`);
                break;
            }
        }
        game.displayHistory();
        const again = await askQuestion('Хотите сыграть ещё раз? (y/n): ');
        if (again.toLowerCase() !== 'y') {
            console.log('До свидания!');
            readline.close();
            process.exit(0);
        }
    }
}

// Аргументы командной строки
if (process.argv.includes('--auto')) {
    const idx = process.argv.indexOf('--auto');
    const diff = parseInt(process.argv[idx+1]) || 2;
    (async () => {
        const game = new Game(diff);
        await game.loadRecord();
        console.log(`Автоигра на уровне ${diff}`);
        let left = game.low, right = game.high;
        while (!game.isGameOver()) {
            const guess = Math.floor((left + right) / 2);
            console.log(`Компьютер пробует: ${guess}`);
            const hint = game.checkGuess(guess);
            game.addAttempt(guess, hint);
            if (game.won) {
                console.log(`Угадал за ${game.attempts} попыток!`);
                break;
            }
            if (hint.includes('▲')) left = guess + 1;
            else right = guess - 1;
            await new Promise(resolve => setTimeout(resolve, 500));
        }
        if (!game.won) console.log('Компьютер не справился :(');
        readline.close();
    })();
} else {
    playGame();
}

File: hotcold.rs
// hotcold.rs - Игра "Горячо-холодно" на Rust
// Для запуска требуется Cargo.toml с зависимостями:
// [dependencies]
// rand = "0.8"
// serde_json = "1.0"

use std::io::{self, Write, BufRead};
use std::fs;
use std::collections::HashMap;
use rand::Rng;
use serde_json;

struct Game {
    difficulty: u8,
    low: i32,
    high: i32,
    max_attempts: i32,
    secret: i32,
    attempts: i32,
    history: Vec<(i32, String)>,
    won: bool,
    record: Option<i32>,
}

impl Game {
    fn new(diff: u8) -> Self {
        let (low, high, max_attempts) = match diff {
            1 => (1, 50, 999),
            2 => (1, 100, 10),
            3 => (1, 200, 7),
            _ => (1, 100, 10),
        };
        let mut rng = rand::thread_rng();
        let secret = rng.gen_range(low..=high);
        let mut game = Self {
            difficulty: diff,
            low,
            high,
            max_attempts,
            secret,
            attempts: 0,
            history: Vec::new(),
            won: false,
            record: None,
        };
        game.load_record();
        game
    }

    fn check_guess(&mut self, guess: i32) -> String {
        if guess == self.secret {
            self.won = true;
            return "ПОБЕДА!".to_string();
        }
        let dist = (guess - self.secret).abs();
        let hint = if dist <= 10 { "ГОРЯЧО" } else if dist <= 25 { "ТЕПЛО" } else { "ХОЛОДНО" };
        let dir = if guess < self.secret { " ▲" } else { " ▼" };
        format!("{}{}", hint, dir)
    }

    fn add_attempt(&mut self, guess: i32, hint: String) {
        self.attempts += 1;
        self.history.push((guess, hint));
    }

    fn is_game_over(&self) -> bool {
        self.won || self.attempts >= self.max_attempts
    }

    fn save_record(&mut self) -> bool {
        let key = format!("{}_{}_{}", self.difficulty, self.low, self.high);
        let mut records: HashMap<String, i32> = HashMap::new();
        if let Ok(data) = fs::read_to_string("records.json") {
            if let Ok(map) = serde_json::from_str(&data) {
                records = map;
            }
        }
        if let Some(&old) = records.get(&key) {
            if self.attempts >= old {
                return false;
            }
        }
        records.insert(key, self.attempts);
        let _ = fs::write("records.json", serde_json::to_string_pretty(&records).unwrap());
        true
    }

    fn load_record(&mut self) {
        let key = format!("{}_{}_{}", self.difficulty, self.low, self.high);
        if let Ok(data) = fs::read_to_string("records.json") {
            if let Ok(records) = serde_json::from_str::<HashMap<String, i32>>(&data) {
                self.record = records.get(&key).copied();
            }
        }
    }

    fn display_history(&self) {
        println!("\x1b[36m--- История попыток ---\x1b[0m");
        for (i, (guess, hint)) in self.history.iter().enumerate() {
            let color = if hint.contains("ГОРЯЧО") { "\x1b[31m" } else { "\x1b[34m" };
            println!("{}: {} -> {}{}\x1b[0m", i+1, guess, color, hint);
        }
        println!("------------------------");
    }
}

fn read_line() -> String {
    let stdin = io::stdin();
    let mut line = String::new();
    stdin.lock().read_line(&mut line).expect("Ошибка ввода");
    line.trim().to_string()
}

fn play_game() {
    println!("\x1b[33m🔥❄️ Добро пожаловать в Hot&Cold!\x1b[0m");
    loop {
        println!("\nВыберите уровень сложности:");
        println!("1. Лёгкий (1-50, безлимит)");
        println!("2. Средний (1-100, 10 попыток)");
        println!("3. Сложный (1-200, 7 попыток)");
        println!("4. Выйти");
        print!("Ваш выбор: ");
        io::stdout().flush().unwrap();
        let choice = read_line();
        if choice == "4" {
            println!("До свидания!");
            break;
        }
        let diff: u8 = match choice.parse() {
            Ok(d) if d >= 1 && d <= 3 => d,
            _ => {
                println!("\x1b[31mНеверный выбор.\x1b[0m");
                continue;
            }
        };

        let mut game = Game::new(diff);

        if let Some(rec) = game.record {
            println!("\x1b[35mТекущий рекорд: {} попыток\x1b[0m", rec);
        } else {
            println!("Рекорда пока нет.");
        }
        println!("\x1b[32mЧисло загадано! Диапазон {}-{}\x1b[0m", game.low, game.high);
        if game.max_attempts < 999 {
            println!("У вас {} попыток.", game.max_attempts);
        } else {
            println!("Попытки не ограничены.");
        }

        while !game.is_game_over() {
            print!("Попытка {}: Введите число: ", game.attempts+1);
            io::stdout().flush().unwrap();
            let input = read_line();
            let guess: i32 = match input.parse() {
                Ok(n) => n,
                Err(_) => {
                    println!("\x1b[31mВведите целое число!\x1b[0m");
                    continue;
                }
            };
            if guess < game.low || guess > game.high {
                println!("\x1b[31mЧисло должно быть от {} до {}\x1b[0m", game.low, game.high);
                continue;
            }
            let hint = game.check_guess(guess);
            game.add_attempt(guess, hint.clone());
            let color = if hint.contains("ГОРЯЧО") { "\x1b[31m" } else if hint.contains("ХОЛОДНО") { "\x1b[34m" } else { "\x1b[32m" };
            println!("{}{}\x1b[0m", color, hint);

            if game.won {
                println!("\x1b[32m🎉 ПОБЕДА! Число {} угадано за {} попыток!\x1b[0m", game.secret, game.attempts);
                if game.save_record() {
                    println!("\x1b[36m🏆 Новый рекорд!\x1b[0m");
                }
                break;
            } else if game.attempts >= game.max_attempts {
                println!("\x1b[31m😞 Вы исчерпали все попытки. Загаданное число: {}\x1b[0m", game.secret);
                break;
            }
        }
        game.display_history();
        print!("Хотите сыграть ещё раз? (y/n): ");
        io::stdout().flush().unwrap();
        let again = read_line();
        if again.to_lowercase() != "y" {
            println!("До свидания!");
            break;
        }
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() > 1 && args[1] == "--auto" {
        let diff: u8 = if args.len() > 2 { args[2].parse().unwrap_or(2) } else { 2 };
        let mut game = Game::new(diff);
        println!("Автоигра на уровне {}", diff);
        let mut left = game.low;
        let mut right = game.high;
        while !game.is_game_over() {
            let guess = (left + right) / 2;
            println!("Компьютер пробует: {}", guess);
            let hint = game.check_guess(guess);
            game.add_attempt(guess, hint.clone());
            if game.won {
                println!("Угадал за {} попыток!", game.attempts);
                break;
            }
            if hint.contains("▲") {
                left = guess + 1;
            } else {
                right = guess - 1;
            }
            std::thread::sleep(std::time::Duration::from_millis(500));
        }
        if !game.won {
            println!("Компьютер не справился :(");
        }
    } else {
        play_game();
    }
}
